### PR TITLE
[libc++] Refactor the predicate taking variant of `__cxx_atomic_wait`

### PR DESCRIPTION
**Pull request by jiixyj**
This is a follow-up PR to <https://github.com/llvm/llvm-project/pull/79265>. It aims to be a gentle refactoring of the `__cxx_atomic_wait` function that takes a predicate.

The key idea here is that this function's signature is changed to look like this (`std::function` used just for clarity):

```c++
__cxx_atomic_wait_fn(Atp*, std::function<bool(Tp &)> poll, memory_order __order);
```

...where `Tp` is the corresponding `value_type` to the atomic variable type `Atp`. The function's semantics are similar to `atomic`s `.wait()`, but instead of having a hardcoded predicate (is the loaded value unequal to `old`?) the predicate is specified explicitly.

The `poll` function may change its argument, and it is very important that if it returns `false`, it leaves its current understanding of the atomic's value in the argument. Internally, `__cxx_atomic_wait_fn` dispatches to two waiting mechanisms, depending on the type of the atomic variable:

1. If the atomic variable can be waited on directly (for example, Linux's futex mechanism only supports waiting on 32 bit long variables), the value of the atomic variable (which `poll` made its decision on) is then given to the underlying system wait function (e.g. futex).
2. If the atomic variable can not be waited on directly, there is a global pool of atomics that are used for this task. The ["eventcount" pattern](<https://gist.github.com/mratsim/04a29bdd98d6295acda4d0677c4d0041>) is employed to make this possible.

The eventcount pattern needs a "monitor" variable which is read before the condition is checked another time. libcxx has the `__libcpp_atomic_monitor` function for this. However, this function only has to be called in case "2", i.e. when the eventcount is actually used. In case "1", the futex is used directly, so the monitor must be the value of the atomic variable that the `poll` function made its decision on to continue blocking. Previously, `__libcpp_atomic_monitor` was _also_ used in case "1". This was the source of the ABA style bug that PR#79265 fixed.

However, the solution in PR#79265 has some disadvantages:

- It exposes internals such as `cxx_contention_t` or the fact that `__libcpp_thread_poll_with_backoff` needs two functions to higher level constructs such as `semaphore`.
- It doesn't prevent consumers calling `__cxx_atomic_wait` in an error prone way, i.e. by providing to it a predicate that doesn't take an argument. This makes ABA style issues more likely to appear.

Now, `__cxx_atomic_wait_fn` takes just _one_ function, which is then transformed into the `poll` and `backoff` callables needed by `__libcpp_thread_poll_with_backoff`.

Aside from the `__cxx_atomic_wait` changes, the only other change is the weakening of the initial atomic load of `semaphore`'s `try_acquire` into `memory_order_relaxed` and the CAS inside the loop is changed from `strong` to `weak`. Both weakenings should be fine, since the CAS is called in a loop, and the "acquire" semantics of `try_acquire` come from the CAS, not from the initial load.

### CODE REVIEW

**Comment by jiixyj** [comment]
As this is based on https://github.com/llvm/llvm-project/pull/79265 , which is not yet merged, I converted this into a draft PR.

File: libcxx/include/__atomic/atomic_sync.h
Comment (ldionne) [inline libcxx/include/__atomic/atomic_sync.h:67]: IIUC, this is roughly what @huixie90 is transforming into a CPO in #81427 .

File: libcxx/include/__atomic/atomic_sync.h
Comment (ldionne) [inline libcxx/include/__atomic/atomic_sync.h:113]: We often use `_fn` to denote things that are implementation details or function objects in the library. IMO, keeping this function named `__cxx_atomic_wait` would be better.

File: libcxx/include/__atomic/atomic_sync.h
Comment (ldionne) [inline libcxx/include/__atomic/atomic_sync.h:139]: I would suggest renaming this to `__atomic_compare_equal_to` or something like that -- that's still not a great name but it is a bit more descriptive, and in the future we could probably write this as a lambda (which is what would really make this code easier to read).

File: libcxx/src/atomic.cpp
Comment (ldionne) [inline libcxx/src/atomic.cpp:181]: Is this because this is handled by `__poll_or_get_monitor` now?

**Review by ldionne — COMMENTED** [review]
Thanks for the patch and sorry it took so long to get to it, the last few weeks have been quite busy. I appreciate that you spent time figuring out how to refactor this code and make it less prone to bugs -- it's important since this code is really complex.

I generally like the approach. I have a few comments and questions but I think we could move forward with this, as it does make the code better and also opens the door to more refactoring like #81427.

I must say I kinda liked the approach suggested by @huixie90 in #79265 where he refactored `latch` and `barrier` to use `atomic::wait` directly, but I guess we can move forward with this patch and then do that other approach in case we end up getting consensus to do it.

I would like @huixie90 to stamp this patch before I give it a last LGTM.

**Comment by huixie90** [comment]
LGTM with comments from Louis resolved.

I could continue the discussion on whether to refactor the code to use public `wait` API after this lands.
also just FYI, I have further refactoring to enable these functions reusable by atomic_ref (which build on top of this patch)
https://github.com/llvm/llvm-project/pull/81427

**Review by ldionne — APPROVED** [review]
Since I am going OOO for a week and I don't want to block this on my approval, please go ahead with the merge once you feel like my comments/questions have been resolved/answered. Don't wait for me to stamp again.

Thanks for the patch, again.

File: libcxx/include/__atomic/atomic_sync.h
Comment (jiixyj) [reply libcxx/include/__atomic/atomic_sync.h:67]: Yes, the two `__poll_or_get_monitor` functions here are basically the same thing as the two `__update_monitor_val_and_poll` functions from #81427 ([here](<https://github.com/llvm/llvm-project/pull/81427/files#diff-7202d31e2a9550950ef2b4630acd4d6d44d9f86e721082c25c0e242a2faccf9fR99-R116>)).

File: libcxx/include/__atomic/atomic_sync.h
Comment (jiixyj) [reply libcxx/include/__atomic/atomic_sync.h]: Yes, a lambda would definitely help. I renamed this struct to `__atomic_compare_unequal_to`. So the logic can be read as "wait unless/until the atomic variable compares unequal to val" (similar to the wording in the spec: <https://eel.is/c++draft/atomics#ref.ops-23.2>).

File: libcxx/include/__atomic/atomic_sync.h
Comment (jiixyj) [reply libcxx/include/__atomic/atomic_sync.h]: > We often use `_fn` to denote things that are implementation details or function objects in the library.

Thanks, that is good to know!

> IMO, keeping this function named `__cxx_atomic_wait` would be better.

I agree. It would conflict with the "normal" `__cxx_atomic_wait` though, as the `_Poll` type parameter is unconstrained. Instead of doing some SFINAE hacks I renamed it to `__cxx_atomic_wait_until` in anticipation of #81427.

File: libcxx/src/atomic.cpp
Comment (jiixyj) [reply libcxx/src/atomic.cpp:181]: Yes, exactly. This overload of `__libcpp_atomic_monitor` was used to get the "monitor" value in case the `ptrdiff_t` of the semaphore could be waited on directly. And this was the reason for the original ABA bug, where this could happen:

1. `__libcpp_atomic_monitor` returns "1" (i.e. the semaphore has the value 1)
2. another thread swoops in, takes the "1" down to "0"
3. "our" poll function returns "false", because there is nothing left to `try_acquire`
4. another thread releases the semaphore (increases the value from "0" to "1")
5. `atomic_wait` now waits on the "1", resulting in a deadlock.

Comment (jiixyj) [reply libcxx/src/atomic.cpp:181]: And now, `__poll_or_get_monitor` makes sure that in this case, the "monitor" value is always "0" (because `return __poll_(__monitor);` sets `__monitor` to "0" if it returns false).

**Comment by jiixyj** [comment]
Thank you @huixie90 and @ldionne for the review! From my side this is good to go.

**Pull request merged**
